Add menu bar icon on macOS

antonio-ramadas opened the pull request:
Fixes #206. Add menu bar icon and implement a global shortcut to toggle the window. The shortcut is Ctrl+M, where M stands for Messenger.

vadimdemedes commented on index.js:
```diff
@@ -37,15 +38,19 @@ if (isAlreadyRunning) {
 	// ignore `Sindre messaged you` blinking
 	if (title.indexOf('Messenger') === -1) {
-		return;
+		return -1;
```
`-1` is usually used for indexes, so I'd return `null` instead.

antonio-ramadas replied:
That makes sense. Done!

vadimdemedes commented on tray.js:
```diff
@@ -37,16 +36,57 @@ exports.create = win => {
 };
 
-exports.setBadge = shouldDisplayUnread => {
+function setBadge(shouldDisplayUnread) {
```
Use arrow functions instead of "regular" ones.

vadimdemedes replied:
`shouldDisplayUnread` => `hasUnreadMessages`.

antonio-ramadas replied:
Seems better to me also. Done!

vadimdemedes commented on index.js:
```diff
@@ -176,7 +181,13 @@ function createMainWindow() {
-		updateBadge(title);
+
+		const messageCount = getMessageCount(title);
+		if (messageCount === -1) {
```
Shouldn't we update the badge anyway? E.g. in case message was just read and the unread count is zero now?

antonio-ramadas replied:
Well, no. This condition is for when the window title is blinking. It is unnecessary to update when it is blinking. The window title is different when received a new message. It is only required to update in this last case.

vadimdemedes commented on tray.js:
```diff
@@ -4,27 +4,26 @@ const electron = require('electron');
-			win.show();
-		}
-	};
+	if (process.platform === 'darwin') {
```
This could be extracted into a `getIconPath()` function, that'd do platform check inside.

antonio-ramadas replied:
You're right. Done!

vadimdemedes commented on tray.js:
```diff
@@ -37,16 +36,57 @@ exports.create = win => {
 		return;
 	}
 
 	const icon = shouldDisplayUnread ? 'IconTrayUnread.png' : 'IconTray.png';
```
Isn't `getDarwinIconPath()` (potentially `getIconPath()` as mentioned above) responsible for determining which icon to display?

antonio-ramadas replied:
The previous refactoring I did now does this. So, done!

vadimdemedes commented on tray.js:
```diff
@@ -37,16 +36,57 @@ exports.create = win => {
+
+	if (counter > 0) {
+		const msg = (counter === 1 ? 'message' : 'messages');
+		tip = tip.concat(' - ', counter, ' unread ', msg);
```
=>

```js
tip += ` - ${counter} unread {msg}`;
```

antonio-ramadas replied:
Complete miss by me. Done!

vadimdemedes commented on tray.js:
```diff
@@ -37,16 +36,57 @@ exports.create = win => {
+	tray.setToolTip(tip);
+}
+
+function getDarwinIconPath(counter) {
```
Since actual number of unread messages (`counter`) isn't needed here, I'd rename it to `hasUnreadMessages` and call `getDarwinIconPath()` like so: `getDarwinIconPath(messageCount > 0)`.

antonio-ramadas replied:
Yes, passing a number may trigger future bugs. Done!

vadimdemedes commented:
Would be useful to be able to hide menubar icon altogether.

vadimdemedes commented:
Thanks for working on this, @antonio-ramadas!

sindresorhus commented on index.js:
```diff
@@ -54,8 +59,8 @@ function updateBadge(title) {
 
-	if (process.platform === 'linux' || process.platform === 'win32') {
-		tray.setBadge(messageCount);
+	if (['linux', 'win32', 'darwin'].indexOf(process.platform) !== -1) {
```
This check is now moot as we only support those systems.

antonio-ramadas replied:
Yes, did not occur to me at the time. Done!

sindresorhus commented:
- Clicking the menubar icon many times doesn't always toggle the window. Sometimes it does nothing.
- The white menubar icon should not have a border.
- It should be possible to hide the menubar icon. Should it even be on by default? I'm not sure. Probably best to have a checkbox menuitem in the "Caprine" menu.

antonio-ramadas commented:
> - Clicking the menubar icon many times doesn't always toggle the window. Sometimes it does nothing.

Yes, I experience the same when I was trying to mess around. However, this only happens when the user toggles too quickly. A normal use works as suppose. This is now the responsibility of a npm module, but I can back it up to our own implementation if you want.

> - The white menubar icon should not have a border.

Ok! Will do when I've some more extra time.

> - It should be possible to hide the menubar icon. Should it even be on by default? I'm not sure. Probably best to have a checkbox menuitem in the "Caprine" menu.

By default, I don't think so. I can hide it and have the checkbox. Thinking about it, if the menubar icon is being shown, then the app may be hidden from the dock. With the global shortcut, when the menubar is enabled (and the dock hidden), the window may appear above all others in the current workspace. I've an app which does this and it is greatly convenient. What do you say?

> - When right-clicking the menubar icon it shows a "Toggle" menu item and "Quit. The "Toggle" menuitem is moot on macOS as left-clicking the icon already toggles the window.

I think this is the same for the other platforms. So, it's probably better to take it out of all platforms.

sindresorhus commented:
> but I can back it up to our own implementation if you want.

👍

sindresorhus commented:
> Thinking about it, if the menubar icon is being shown, then the app may be hidden from the dock. With the global shortcut, when the menubar is enabled (and the dock hidden), the window may appear above all others in the current workspace. I've an app which does this and it is greatly convenient. What do you say?

Agreed. I think this is the usual behavior for "menubar mode" in GUI apps.

antonio-ramadas commented:
- Added the double-click event to the menubar icon to be able to toggle more effectively.
- Implemented our own toggle window implementation.
- Wrapped up the tray function into a _continuity mode_ which only works on macOS. This mode registers the global shortcut, places the window on top of everything, fades when lost focus, the dock icon disappears and the menubar icon (tray) appears. Disabling it returns to _normal mode_.

Finally, @sindresorhus I believe I didn't understand what you said about the borders about the white icon. When I removed the thicker black line, it only appeared the lightning bolt with no background. Is this what you want?

antonio-ramadas commented:
Also, I experienced random crashes when disabling the _continuity mode_. Sometimes the app crashed on the first time, but there were times I had to _play_ for some time switching between modes until I crashed it. I've spent some time tracing it, but I'm not finding the bug. Also, Electron exits with code 0 and shows:
`Received signal 11 SEGV_MAPERR 000000001940`

Any help is appreciated!

sindresorhus commented:
If I choose "Disable Continuity" while the Caprine window is hidden, it should be shown again.

---

Clicking the Caprine menu bar icon fast stops it from being toggled.

---

> Finally, [@sindresorhus](https://github.com/sindresorhus) I believe I didn't understand what you said about the borders about the white icon. When I removed the thicker black line, it only appeared the lightning bolt with no background. Is this what you want?

Notice the difference in the Caprine icon and another one:

<img width="730" alt="screen shot 2017-07-23 at 15 18 49" src="https://user-images.githubusercontent.com/170270/28499655-074b9bb2-6fbb-11e7-9451-5842c872e8e6.png">
<img width="610" alt="screen shot 2017-07-23 at 15 19 27" src="https://user-images.githubusercontent.com/170270/28499656-074bf896-6fbb-11e7-93eb-772885fb8fd2.png">

The black outer border should not be there.

And the lighting icon should be transparent, so it would be blue when highlighted.

// @mirshko

---

Can you add this feature to the https://github.com/sindresorhus/caprine#highlights section and https://github.com/sindresorhus/caprine#features

And add the keyboard shortcut to https://github.com/sindresorhus/caprine#keyboard-shortcuts

---

Why is it named "Continuity"?

mirshko commented:
@sindresorhus this is just a white version of the unread black logo. Is that what you want?

[unread_1.svg.zip](https://github.com/sindresorhus/caprine/files/1170216/unread_1.svg.zip)

sindresorhus commented:
The current version (shown above) is using a filled icon with border. It should either be the border without fill or filled without a border, like previewed in https://github.com/sindresorhus/caprine/issues/206#issuecomment-307985319. I think the border one looks the best.

mirshko commented:
[Archive.zip](https://github.com/sindresorhus/caprine/files/1171107/Archive.zip)

Okay these should be what you need. I think I understand now.

antonio-ramadas commented:
> If I choose "Disable Continuity" while the Caprine window is hidden, it should be shown again.

Done

> Clicking the Caprine menu bar icon fast stops it from being toggled.

I believe I know what you're talking about, but I'm not figuring out  a solution for it. I created a dummy app just (https://github.com/antonio-ramadas/electron-tray) to test if it was caused by Caprine. I concluded that there is not correlation.

> Menu bar icon

Done

> README.md

Done

> Why is it named "Continuity"?

While working I sometimes need to type quick responses, but I do not like switching context. That's why I'm doing this. This reminds me Apple's [continuity](https://support.apple.com/en-gb/HT204681) as the switch is seamless.

antonio-ramadas commented:
While working on this, I found a bug. While the continuity is enabled, if the window is maximised, you can't minimize it easily because the menu bar does not show. I can think of 3 solutions while continuity is enabled:

1 - Prevent maximize

2 - Show the app on the dock when the window is maximised (which will enable the menu bar)

3 - Add a button to minimize the window

sindresorhus commented:
@antonio-ramadas I'm getting an error:

<img width="532" alt="screen shot 2017-09-29 at 13 51 24" src="https://user-images.githubusercontent.com/170270/31004293-fd92488a-a51d-11e7-9731-8aa6abbf6ea3.png">

antonio-ramadas commented:
Seems that I forgot to upload the icons. I'll take care of that.

antonio-ramadas commented:
@sindresorhus Done!

sindresorhus commented:
> While working I sometimes need to type quick responses, but I do not like switching context. That's why I'm doing this. This reminds me Apple's continuity as the switch is seamless.

I think users will confuse this with the actual system Continuity thing, which is totally different from this. Can we just be explicit an call it `Menu Bar Mode`?

> 1 - Prevent maximize

👍

sindresorhus commented:
The icon is way to large now. Compare it to other menu bar icons:

<img width="151" alt="screen shot 2017-11-20 at 17 47 36" src="https://user-images.githubusercontent.com/170270/33015026-e06c93ba-ce1b-11e7-96f5-ab35476247d7.png">

AllanLRH commented:
Any progress of this? Would really like the menu-bar functionality, so that I (hopefully) can hide the app from the app-switcher by adding 
```
<key>LSUIElement</key>
<true/>
```
to `/Applications/Caprine.app/Contents/Info.plist`, yet still access the app when needed.

Currently I just start the app when I need to write to someone, or my phone notifies me that someone are writing to me :)

antonio-ramadas commented:
@sindresorhus I'll take care of that when I've some more time available.

@AllanLRH Any help is welcome. Anyway, according to the latest changes requested, this pull request seems to be close to being accepted.

sindresorhus commented:
Closing in favor of #374
